lib/Conversion/RTLToLLHD/RTLToLLHD.cpp

```diff
@@ -207,39 +272,54 @@ struct ConvertInstance : public OpConversionPattern<InstanceOp> {
 
       Location loc = result.getLoc();
 
-      Value init;
+      // Since we need to have a signal for this result, see if an OutputOp maps
+      // it to an output signal of our parent module. In that case we can just
+      // use that signal.
+      Value sig;
       for (auto &use : result.getUses()) {
-        Value sig;
-
         if (isa<OutputOp>(use.getOwner())) {
-          // If the result is an entity output, use that signal.
           auto entity = instance->getParentOfType<EntityOp>();
+          if (!entity)
+            continue;
           sig = entity.getArgument(entity.ins() + use.getOperandNumber());
-        } else {
-          // Otherwise, materialize a signal.
-          if (!init)
-            init = rewriter.create<ConstOp>(
-                loc, resultType, rewriter.getIntegerAttr(resultType, 0));
-
-          SmallString<8> sigName(instance.instanceName());
-          sigName += "_result_";
-          sigName += std::to_string(result.getResultNumber());
-          sig = rewriter.createOrFold<SigOp>(loc, SigType::get(resultType),
-                                             sigName, init);
+          break;
         }
+      }
 
-        // Replace all uses of this result with the signal.
-        rewriter.updateRootInPlace(use.getOwner(), [&]() { use.set(sig); });
+      // Otherwise materialize a signal.
+      // TODO: This should be a `llhd.buffer` operation.
+      if (!sig) {
+        auto init = rewriter.create<ConstOp>(
+            loc, resultType, rewriter.getIntegerAttr(resultType, 0));
+        SmallString<8> sigName(instance.instanceName());
+        sigName += "_result_";
+        sigName += std::to_string(result.getResultNumber());
+        sig = rewriter.createOrFold<SigOp>(loc, SigType::get(resultType),
+                                           sigName, init);
+      }
```

Comment:
@mikeurbach I think you implemented what this snippet does with another materialization. I wonder if we could just move this code into the materialization callback to get nice `sig`+`drv` combinations. I'm a bit worried that this won't be possible to do without canonicalization cleaning up `prb`+`sig`+`drv` afterwards, which is tricky in LLHD.

Reply:
Yes, I think this is sort of what this target materialization was for: https://github.com/llvm/circt/pull/990/files#diff-8460c806a460e177b99798deab2307f3796bbd268449b25e1d470add36157ef7R101. It's kind of hard to know exactly the signature needed for a materialization, but in this case the dialect conversion debug logs will say something about being unable to materialize a conversion for operand #n.

In general, doing target materializations that generate a signal seem like they'd really benefit from the hypothetical `llhd.buffer` op. As you say, it is a little awkward to materialize signals right now, and we kind of need something that plays nice in terms of MLIR SSA use-def but also has the right semantics for LLHD.

Reply:
What's bothering me about this is that I'm having a hard time figuring out how we can collapse `prb`+`drv` in LLHD: the drive is always associated with a delay (and so would a hypothetical `llhd.buffer`). There's a semantic difference between applying the value of one net to another with a delay, compared to declaring two nets equal with a connect. But I see the need for being able to reason and deduct the latter from a `prb`-`drv` chain. The delays and subtle differences between directed buffering/driving from one signal to another and the bidirectional coupling of signals with a connect are a nightmare.

Reply:
This may be naive, so forgive me as I am still learning about LLHD's timing model. Is it possible to have a time with 0 delay? I tried creating an attribute like `#llhd.time<0ns, 0d, 0e>` and nothing seemed to complain, at least in parsing. Not sure if the ops verifiers' would reject it.

Maybe we can check if there is 0 delay on a prb - drv chain to determine if two nets can just be connected?

Reply:
The 0 delay route is definitely a nice idea. I'm not sure how this should behave in simulation though. When two wires are connected together with `llhd.con`, you expect the nets to become equal, with drives to one net immediately appearing on the other, *and* vice versa. It's like a short-circuiting with a copper wire. The LLHD event queue (and the SV/VHDL one which it borrows from) expect drives to signals to only appear after some delay. This holds even for blocking assignments in SV, which weirld appear to take effect immediately, but actually don't really and can still cause a delta step (the spec is horribly vague). In this model we'd need to figure out what it means to drive with 0 delay: does the event loop run twice for the same time step? Do we expand 0 delay drives to a delta-step in a pre-pass (and who does that change the meaning)? Or do we say 0 delay drives have a special meaning in that their logic cone is gobbled up into the event queue itself, and runs as events are being applied to signals?

In any case, I think we need a uni-directional connect to be able to fold `%0 = llhd.prb %a; llhd.drv %b, %0` to `llhd.connect_oneway %a, %b`. And we need that `llhd.buffer` to do a nice value-to-signal conversion (which is what you want in `llhd.entity` in general).

Reply:
Coming to think of it: defining clear semantics for 0 delay in the LLHD event loop is a good idea. That could help flesh out what `llhd.con` actually does. Right now connects are not really implemented, and they have extremely ugly effects on how events get processed. Essentially the connect declares two signals to behave as if they were one. But that gets really weird if you shift a signal with by a dynamic value, and then connect it to something else. Consider the following SV:

```systemverilog
module Foo(input [2:0] addr, inout [31:0] io);
  bit [7:0][31:0] mem;
  Bar dummy(mem[addr], io); // connects the two arguments
endmodule

// The following is a hack to truly connect the first and second port argument.
module Bar(a,a);
  inout bit [31:0] a;
endmodule
```
This would translate to roughly the following LLHD (don't remember the vector syntax off the top of my head), with the `Bar dummy` instance inlined:
```mlir
llhd.entity @Foo (%addr: !llhd.sig<i3>) -> (%io: !llhd.sig<i32>) {
  %0 = [...] : !llhd.vector<8 x i32>
  %mem = llhd.sig "mem" %0 : !llhd.sig<vector<8 x i32>>
  %1 = llhd.prb %addr : i3
  %2 = llhd.dshr %mem, %1 : (!llhd.sig<vector<8 x i32>>, i3) -> !llhd.sig<i32>
  // inlined `Bar dummy` instance:
  llhd.con %2, %io : !llhd.sig<i32>
}
```
I have no idea what the precise semantics of driving a value to `io` from outside `Foo` are. I'm pretty sure that according to SV this is well-defined. My guess is that scheduling a drive on `io` will also apply to the `mem` index selected by the `addr` at the very moment the event is applied. But I'm also pretty sure that many commercial tools will disagree on this :wink:.

I really dislike these lvalue-style signals in LLHD, but they are common enough in SV/VHDL that we need to deal with them :thinking:

Reply:
These are good points, and I'm intrigued to see how far this 0-delay concept might go. I was imagining 0-delay drives might be present in the IR through up until some point that they are resolved to connections, turned into a delta step, etc. (or a pass gives up and errors out). I.e. they might be usefuly for analysis in some intermediate forms, but potentially need to be dealt with before simulation.

Another thought that your comment about lvalues brings up--it is sort of how the RTL dialect has InOutType. That is an ugly reality, and in both the type system and the ops, the RTL dialect has worked really hard to cordon it off (for example by kicking out rtl.wire and rtl.reg to the SV dialect) and handle it separately from "everything else". So I can imagine a world where most of LLHD works nicely with "rvalues", the conversion here is super smooth with simple, obvious materializations, and ops like probe and drive are only for special cases.

Anyway, this has been a really interesting issue for me, and I think we are on the same page about the best way to use the dialect conversion framework in practice. It's up to you if you want to keep discussing/designing/implementing these things, or just move forward for now with the existing ops and come back to this later.

Reply:
Likewise! I'd suggest to merge this as a first workign state, then follow exactly your steps and properly define drives with no delay, a buffer instruction for SSA-style value-to-signal conversion, and then further clean this lowering up to leverage the new operations in a future PR.

Reply:
Great, I will give it a solid review.